File: L4_Expert/BEAMhard/Kimi K3/serve.py
#!/usr/bin/env python3
"""Tiny static server for the THW_CCF2 web app. Usage: py serve.py [port]"""
import http.server, socketserver, sys, os

PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 8000
os.chdir(os.path.dirname(os.path.abspath(__file__)))

class Handler(http.server.SimpleHTTPRequestHandler):
    extensions_map = {
        **http.server.SimpleHTTPRequestHandler.extensions_map,
        '.js': 'text/javascript', '.mjs': 'text/javascript',
        '.jbeam': 'text/plain', '.dae': 'text/xml', '.dds': 'application/octet-stream',
        '.json': 'application/json',
    }
    def log_message(self, fmt, *args):
        pass

with socketserver.ThreadingTCPServer(("", PORT), Handler) as httpd:
    httpd.allow_reuse_address = True
    print(f"Serving on http://localhost:{PORT}  (Ctrl+C to stop)")
    httpd.serve_forever()

File: L4_Expert/BEAMhard/Opus/scripts/serve.mjs
#!/usr/bin/env node
/**
 * Zero-dependency static file server for the CCF WebGL vehicle lab.
 * Usage:  node scripts/serve.mjs [port]
 * Serves the repository root with correct MIME types for .dae/.dds/.jbeam.
 */
import http from 'node:http';
import { createReadStream, statSync } from 'node:fs';
import { extname, join, normalize, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = resolve(fileURLToPath(new URL('..', import.meta.url)));
const PORT = Number(process.argv[2] || process.env.PORT || 8080);

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.jbeam': 'application/json; charset=utf-8',
  '.pc': 'application/json; charset=utf-8',
  '.dae': 'model/vnd.collada+xml',
  '.dds': 'image/vnd-ms.dds',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.svg': 'image/svg+xml',
  '.wav': 'audio/wav',
  '.ogg': 'audio/ogg',
  '.ini': 'text/plain; charset=utf-8',
  '.md': 'text/markdown; charset=utf-8',
};

const server = http.createServer((req, res) => {
  try {
    const url = new URL(req.url, `http://${req.headers.host}`);
    let pathname = decodeURIComponent(url.pathname);
    if (pathname.endsWith('/')) pathname += 'index.html';
    const filePath = join(ROOT, normalize(pathname).replace(/^(\.\.[/\\])+/, ''));
    if (!filePath.startsWith(ROOT)) { res.writeHead(403); res.end('forbidden'); return; }
    let st;
    try { st = statSync(filePath); } catch { res.writeHead(404); res.end('not found: ' + pathname); return; }
    if (st.isDirectory()) { res.writeHead(302, { Location: pathname + '/' }); res.end(); return; }
    res.writeHead(200, {
      'Content-Type': MIME[extname(filePath).toLowerCase()] || 'application/octet-stream',
      'Content-Length': st.size,
      'Cache-Control': 'no-cache',
      'Cross-Origin-Opener-Policy': 'same-origin',
      'Cross-Origin-Embedder-Policy': 'require-corp',
    });
    createReadStream(filePath).pipe(res);
  } catch (err) {
    res.writeHead(500); res.end(String(err));
  }
});

server.listen(PORT, () => {
  console.log(`[serve] CCF Vehicle Lab at http://localhost:${PORT}/  (root: ${ROOT})`);
});
